zcash_client_sqlite/src/lib.rs

```diff
@@ -840,4 +894,92 @@ mod tests {
+        use zcash_primitives::zip32::ExtendedSpendingKey;
+
+        use crate::{
+            chain::init::init_cache_database,
+            wallet::{get_balance, init::init_wallet_db, rewind_to_height},
+        };
+
+        let cache_file = NamedTempFile::new().unwrap();
+        let db_cache = BlockDb::for_path(cache_file.path()).unwrap();
+        init_cache_database(&db_cache).unwrap();
+
+        let data_file = NamedTempFile::new().unwrap();
+        let mut db_data = WalletDb::for_path(data_file.path(), network()).unwrap();
+        init_wallet_db(&mut db_data, Some(Secret::new(vec![]))).unwrap();
+
+        // Add an account to the wallet.
+        let (dfvk, _taddr) = init_test_accounts_table(&db_data);
+
+        // Account balance should be zero.
+        assert_eq!(
+            get_balance(&db_data, AccountId::from(0)).unwrap(),
+            Amount::zero()
+        );
+
+        // Create a fake CompactBlock sending value to the address.
+        let value = Amount::from_u64(5).unwrap();
+        let (cb, nf) = fake_compact_block(
+            sapling_activation_height(),
+            BlockHash([0; 32]),
+            &dfvk,
+            value,
+        );
+        insert_into_cache(&db_cache, &cb);
+
+        // Create a second fake CompactBlock spending value from the address.
+        let extsk2 = ExtendedSpendingKey::master(&[0]);
+        let to2 = extsk2.default_address().1;
+        let value2 = Amount::from_u64(2).unwrap();
+        let cb2 = fake_compact_block_spending(
+            sapling_activation_height() + 1,
+            cb.hash(),
+            (nf, value),
+            &dfvk,
+            to2,
+            value2,
+        );
+        insert_into_cache(&db_cache, &cb2);
+
+        // Scan the cache.
+        let mut db_write = db_data.get_update_ops().unwrap();
+        scan_cached_blocks(&network(), &db_cache, &mut db_write, None).unwrap();
+
+        // Account balance should equal the change.
+        assert_eq!(
+            get_balance(&db_data, AccountId::from(0)).unwrap(),
+            (value - value2).unwrap()
+        );
+
+        // Remove the second transaction.
+        let mut db_ops = db_data.get_update_ops().unwrap();
+        let txid2 = cb2.vtx.get(0).unwrap().txid();
+        db_ops.remove_tx(&txid2).unwrap();
+
+        // Account balance should only reflect the first received note.
+        assert_eq!(get_balance(&db_data, AccountId::from(0)).unwrap(), value);
+
+        // Scanning the cache again won't change the balance, as the wallet believes it
+        // has already scanned all the blocks in the cache.
+        scan_cached_blocks(&network(), &db_cache, &mut db_write, None).unwrap();
+        assert_eq!(get_balance(&db_data, AccountId::from(0)).unwrap(), value);
+
+        // Rewind the wallet by one block, and scan the cache again.
+        rewind_to_height(&db_data, sapling_activation_height()).unwrap();
```

Comment:
This inconsistency in database state bothers me a little bit. It really seems like we shouldn't be exposing the ability to remove a transaction directly, but should instead be manipulating the chain state to get rid of the transaction so that the two don't get out of sync.

Reply:
The Android SDK code that wants this, is wanting to manipulate _non-chain_ state: either just-created transactions that haven't been mined, or transactions that were created and then expired before getting mined. So there's no chain state to keep in sync from that perspective. However, see my other comment where I identify a bug in the upstream logic that this PR preserves around mined transactions.

Reply:
For clarity, I'm leveraging in this test the fact that it is equally fine to remove a transaction that is unmined, and has _just_ been mined, in order to make defining the test easy. The bug is that the Android code (and this code by extension) allows long-mined transactions to also be removed.

Reply:
We now only allow unmined transactions to be removed.